django/db/backends/base/operations.py

```diff
@@ -526,30 +526,40 @@ def adapt_ipaddressfield_value(self, value):
         """
         return value or None
 
-    def year_lookup_bounds_for_date_field(self, value):
+    def year_lookup_bounds_for_date_field(self, value, iso_year=False):
         """
         Return a two-elements list with the lower and upper bound to be used
         with a BETWEEN operator to query a DateField value using a year
         lookup.
 
         `value` is an int, containing the looked-up year.
+        `iso_year` is a boolean, to use isocalendar instead of 1.1. and 31.12.
         """
-        first = datetime.date(value, 1, 1)
-        second = datetime.date(value, 12, 31)
+        if iso_year:
+            first = datetime.date.fromisocalendar(value, 1, 1)
+            second = datetime.date.fromisocalendar(value + 1, 1, 1) - datetime.timedelta(days=1)
+        else:
+            first = datetime.date(value, 1, 1)
+            second = datetime.date(value, 12, 31)
         first = self.adapt_datefield_value(first)
         second = self.adapt_datefield_value(second)
         return [first, second]
 
-    def year_lookup_bounds_for_datetime_field(self, value):
+    def year_lookup_bounds_for_datetime_field(self, value, iso_year=False):
```

Comment:
We could add a small release notes about this change in the [Database backend API](https://docs.djangoproject.com/en/dev/releases/4.0/#database-backend-api) section, e.g.:
```
* ``DatabaseOperations.year_lookup_bounds_for_date_field()`` and
  ``year_lookup_bounds_for_datetime_field()`` now take the optional
  ``iso_year`` argument in order to support bounds for ISO-8601 week-numbering
  years.
```

Reply:
done: bd0de1176c02a895ec88db6a3693ccfbb6427aca